### DISCUSSION TITLE
When is Android support expected?

### DISCUSSION THREAD

**Discussion by roke-julian-lockhart** [twinbasic/twinbasic#1617] [Q&A] [answered]
Per [`twinbasic.com/preorder.html`](https://twinbasic.com/preorder.html#:~:text=planned%20for%20Linux%2C-,mac%2C%20android\),-Additional%20private%20product), Android support is purportedly expected. I'm incredibly surprised by this, because that's not *necessarily* a simple task.

My knowledge of this project comes from a few years ago, and back then, it appeared to use VSCode for its IDE. Is that true? VSCode is supported well as a PWA in most respects (try [`insiders.vscode.dev`](https://insiders.vscode.dev/)) but its text manipulation features are lacklustre on Android, and OS integration is poor owing to the fact that any kind of usable terminal would require Termux integration, probably using Shizuku.

**Comment by fafalone** [comment]
The most up to date estimates for Android support can be found [on the roadmap](https://github.com/twinbasic/twinbasic/issues/335).

twinBASIC now has it's own WebView2-based IDE rather than using VSCode, owing to running up against some limitations of that IDE, but it still uses the same Monaco editor VSCode is built around. 

Btw-- There's been _enormous_ progress since the days of the VSCode plugin; if that's your last reference point for the project it's certainly worth taking another look at where it's at today; the most recent version can always be found in the [Releases section](https://github.com/twinbasic/twinbasic/releases) of this repo. For instance, tB can now use in-project UserControls, all basic controls are available, including data binding, as well as a number of the common controls, and it runs large, complex UCs like my [ucShellBrowse](https://github.com/fafalone/ShellControls) and Krool's [VBFlexGrid](https://github.com/Kr00l/VBFLXGRD) which have both been updated for 64bit support. Also worth checking out my updated version of Carles PV's [Lemmings game](https://github.com/fafalone/Lems64) and the [basicNES NES emulator](https://github.com/fafalone/MiscPorts).

**Reply by roke-julian-lockhart** [reply, marked as answer]
Thanks for that, @fafalone. Seems that the answer is [`issues/335`](https://github.com/twinbasic/twinbasic/issues/335#issue-963668370:~:text=2025%2DQ4%20(OCT%2DDEC),to%20paid%20Ultimate%20Edition%20subscribers\)):

> Start cross-platform support for ARM Linux/Mac/Android ([only available to paid Ultimate Edition subscribers](https://www.twinbasic.com/preorder.html)).

**Reply by roke-julian-lockhart** [reply]
> it's certainly worth taking another look at where it's at today

It doesn't install with Wine since WebView2 doesn't, so I'll have to wait until then to use it. Only got Linux here.

**Comment by mansellan** [comment]
Are you talking about targeting Linux/Android, or developing on them? I expect both of those will land at roughly the same time, but they're not the same thing.

**Reply by roke-julian-lockhart** [reply]
@mansellan, developing on them. I hadn't even considered that anything compatible with VB6 would be able to target Linux, much less Android! Shall or does TwinBasic ultimately delegate its code to a kind of JVM-like framework?

**Reply by fafalone** [reply]
Oh, yes, that's what the roadmap entries are referring to, making applications for them. I believe the plan is for native code; obviously this cannot support everything from VB6. But it can support the same syntax. The Forms/Controls will have to use an alternative framework or be custom written as in VB6 and tB for Windows they're built on the Windows common controls. So you could make something that looks like a Command Button, supports the same built-in methods, but wouldn't for example support `SendMessage(hwnd, BM_SETIMAGE, ...)` because it's not a true Windows control.

**Reply by mansellan** [reply]
It's definitely going to be native, twinBASIC doesn't do runtime VMs.

**Reply by mansellan** [reply]
Except maybe WebAssembly. But that's entirely stack-based, so pretty close to the metal. No managed garbage collection.
